### ISSUE TITLE
Enumerable.Cast<TResult> () should result in a compiler error for types which are not relevant to IConvertible or TResult

### ISSUE THREAD

**Issue by johndog** [dotnet/roslyn#15133]
**Version Used**: 4.5

**Steps to Reproduce**:

Where SomeType does not implement IConvertible:

1. new List<string>(new [] {"StringRepresentation"}).Cast<SomeType>();

Or:

2. new List<SomeType>(new [] {new SomeType()}).Cast<string>();

**Expected Behavior**:  Due to appropriate constraints that do not yet exist on the Cast<> generic, the compiler should signal an error in both cases.  In case 1, the error should be that SomeType() is not one of the types that IConvertible can convert to, and nothing can derive directly from string.  In case 2, the error should be that SomeType does not implement IConvertible, so that it's not possible for it to be converted to a string through a direct cast.

**Actual Behavior**:  This use of Cast<> doesn't get exposed until runtime, when an InvalidCastException() is thrown.  Given that all the types are known statistically, this seems avoidable.

**Comment by CyrusNajmabadi** [comment]
> Due to appropriate constraints on the Cast<> generic, the compiler should signal an error in both cases.

Which constraints are you referring to?  The source for 'Cast' is here: https://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,908

```c#
public static IEnumerable<TResult> Cast<TResult>(this IEnumerable source)
```

**Comment by Joe4evr** [comment]
The problem with this is the following:
```cs
public class Foo
{
    //....
}

public class Bar : Foo
{
    //....
}
```
Suppose you get an `IEnumerable<Foo>` that you happen to know are all instances of `Bar` and you need one of `Bar`'s members? `.Cast()` is just a shorthand for `.Select(foo => (Bar)foo)`, there's no `IConvertable` in play here.

**Comment by johndog** [comment]
@CyrusNajmabadi - Correct, there are no such constraints.  I don't believe those constraints are currently expressible in the language.

It would have to be something like:

```
public static IEnumerable<TElement, TResult> Cast<TResult>(this IEnumerable<TElement> source) 
where (TElement is primitive AND TResult is castable to primitive) OR (TElement is IConvertible AND TResult is primitive) OR (TElement is TResult) OR (TElement is castable to TResult)
```

If this whole thing was expressible, I'm sure it could be collapsed a bit.  But the ability to express and enforce any aspect of this would be an improvement.  Currently there is virtually no compile-time validation of Cast<>.

**Comment by johndog** [comment]
@Joe4evr - That's fair, I've updated the title.  However, one thing you wrote isn't quite correct, and it's part of the problem: `.Cast()` is not actually shorthand for `.Select(foo => (Bar)foo)`.  Consider an alteration of your example, with conversion operators:
```
public class Foo
{
    public static implicit operator Bar(Foo foo)
    {
        return this.bar;
    }
    //....
}

public class Bar
{
    public static explicit operator Foo(Bar bar)
    {
        return new Foo(bar);
    }
    //....
}
```
This is a case where `.Select(foo => (Bar)foo)` works and .Select(bar => (Foo)bar) works, and they work so transparently that you would be fooled into thinking Cast<Bar>() and Cast<Foo>() also works.  But instead you get InvalidCastException.  These conversions are a common method of transitioning from patterns of inheritance to patterns of composition, so it can be alarming to see these uses of Cast<> very quietly and insidiously turning into runtime exceptions.

I included in my response to Cyrus some hypothetical language to include this in the constraint.  But I suspect this is a harder problem and full of nuance, so I would be content if the use of Cast<> in the context of a conversion operator merely resulted in a compiler error instead of an InvalidCastException.

**Comment by svick** [comment]
> I included in my response to Cyrus some hypothetical language to include this in the constraint.

Those constraints do no make sense to me in the context of the current version of `Cast`:

* `TElement is primitive AND TResult is castable to primitive`: `Cast` cannot be used for that kind of cast today
* `TElement is IConvertible AND TResult is primitive`: `Cast` has no relation to `IConvertible` (as has already been pointed out by @Joe4evr)
* `TElement is TResult`: `Cast` does work in this situation, but it also works in other cases too (like `TResult is TElement`, but also others)
* `TElement is castable to TResult`: I have no idea what is `castable` supposed to mean

Are you proposing to extend `Cast` to work in all these cases?

> But I suspect this is a harder problem and full of nuance, so I would be content if the use of Cast<> in the context of a conversion operator merely resulted in a compiler error instead of an InvalidCastException.

The compiler currently does not have any special knowledge about `Cast` and I seriously doubt it would be acceptable to add such knowledge.

Also, doing that would be a breaking change (since bad `Cast` works on zero-element enumerables and you can also catch ICE), so I think such change wouldn't be accepted for that reason alone, even if it was simple to implement (which it isn't).

**Comment by johndog** [comment]
@svick - No, I clearly got my wires crossed.  

I encountered [InvalidCastException ](https://msdn.microsoft.com/en-us/library/system.invalidcastexception(v=vs.110).aspx) in the use of `Cast<>`.  In reading about the exception on MSDN, I learned about its relationship to IConvertible, and must have concluded from this that `Cast<>` supports IConvertible.  My mistake.

So what I want has nothing to do with IConvertible, and comes down entirely to what I wrote to Joe4evr. I think I have to close this and open a new one.
